scipy/signal/tests/test_signaltools.py

```diff
@@ -156,41 +156,49 @@ def test_invalid_shapes(self):
-                            or np.dtype(dtype).name in {'complex256', 'complex192'}:
-                        assert_equal(choose_conv_method(x, h, mode=mode), 'direct')
-                        # continues the for-loop; it will throw an error
-                        # because fftconv doesn't support the dtype
-                        continue
-
-                    if x.dtype.kind != 'b':
-                        x += 1
-                        h += 1
-
-                    if x.dtype.kind == 'c':
-                        x += 1j * np.random.rand(*x.shape).astype(dtype)
-                        h += 1j * np.random.rand(*h.shape).astype(dtype)
-
-                    results = {'direct':convolve(x, h, mode=mode,
-                                                 method='direct'),
-                               'fft':convolve(x, h, mode=mode, method='fft')}
-
-                    rtol = {'rtol': 2.0e-3} if dtype in {np.complex64,
-                                                         np.float32} else {}
-                    if isinstance(results['direct'], np.ndarray):
-                        assert_allclose(results['fft'], results['direct'], **rtol)
-                        assert_equal(results['direct'].dtype, results['fft'].dtype)
+    def test_convolve_method(self, n=100):
+        types = sum([t for _, t in np.sctypes.items()], [])
+        types = {np.dtype(t).name for t in types}
+
+        # These types include 'bool' and all precisions (int8, float32, etc)
+        # The removed types throw errors in correlate or fftconvolve
+        for dtype in ['complex256', 'float128', 'str', 'void', 'bytes',
+                      'object', 'unicode', 'string']:
+            if dtype in types:
+                types.remove(dtype)
+
+        args = [(t1, t2, mode) for t1 in types for t2 in types
+                               for mode in ['valid', 'full', 'same']]
+
+        array_types = {'i': np.random.choice([0, 1], size=n),
+                       'f': np.random.randn(n)}
+        array_types['b'] = array_types['u'] = array_types['i']
+        array_types['c'] = array_types['f'] + 0.5j * array_types['f']
+
+        for t1, t2, mode in args:
+            np.random.seed(42)
+            x1 = array_types[np.dtype(t1).kind].astype(t1)
+            x2 = array_types[np.dtype(t2).kind].astype(t2)
+
+            results = {key: convolve(x1, x2, method=key, mode=mode)
+                       for key in ['fft', 'direct']}
+
+            assert_equal(results['fft'].dtype, results['direct'].dtype)
+
+            if 'bool' in t1 and 'bool' in t2:
+                assert_equal(choose_conv_method(x1, x2), 'direct')
+                continue
+
+            if any([t in {'complex64', 'float32'} for t in [t1, t2]]):
+                kwargs = {'rtol': 8.0e-3, 'atol': 1e-3}
+            elif 'float16' in [t1, t2]:
+                kwargs = {'rtol': 1e-2, 'atol': 1e-5}
+            else:
+                kwargs = {'rtol': 1e-5, 'atol': 1e-5}
```

Comment:
Can you add a few comments on how you determined these tolerance levels for the different types?